<font size="5"><b>Kementerian Keuangan RI </b></font><br/>
<font size="4"><b>Badan Pendidikan dan Pelatihan Keuangan<b/></font><br/>
<font size="4"><b>Pusdiklat Keuangan Umum<b/></font><br/>
### Praktikum Data Scientist Specialist

***
### Perhatikan petunjuk pengisian berikut :

1. Tujuan praktikum ini adalah untuk mengukur kemampuan Anda dalam melakukan pengolahan data dan mengaplikasikan metode-metode machine learning.
2. Anda diperbolehkan melihat file <u>jupyter-notebook yang telah disediakan sebelumnya atau menggunakan search engine </u> untuk menyelesaikan soal-soal yang disediakan.
3. Perhatikan petunjuk pengisian pada masing-masing bagian. Tambahan pekerjaan (step) untuk pengolahan data akan memberikan nilai bonus.
4. <font color="red"><b>Waktu menyelesaikan</b></font> seluruh soal berikut adalah <font color="red"><b>60 menit</b></font>.
***

Tuliskan Nama dan NIP Anda :





***

## Data Science Story Project
> Anda diminta oleh Kementerian Kelautan Republik sebelah sana untuk membangun model yang akan digunakan untuk mengklasifikasikan suatu spesies ikan dari berat, dan ukuran suatu ikan. 

> Anda diminta untuk membuat model klasifikasi yang paling akurat dari beberapa model yang ada, **paling sedikit memuat 3 model** antara lain yaitu **Logistic Regression, RandomForest, Support Vector Machine, dan/atau Multi Layer Perceptron**,. 

> Anda juga diminta untuk memberikan perbandingan model berdasarkan akurasi.


Terdapat tahapan pada jupyter notebook yang sudah dilengkapi secara default, namun Anda diminta untuk tetap melakukan tahapan-tahapan dasar standard *machine learning* yaitu *exploratory data analysis* sebelum membangun model.
### Import library dan load dataset

**1. Hal yang pertama kali harus dilakukan adalah meng-import *library* pandas dan numpy.**

In [1]:
# import libraries 
import pandas as pd
import numpy as np

Selanjutnya, kita akan mengimport (load) data dalam bentuk file `csv` ke dalam objek bernama `df` dengan menggunakan *method* `read_csv`. 

In [2]:
df = pd.read_csv('Fish.csv')

### I. Exploratory Data Analysis
Pada tahapan ini Anda diminta untuk meng-explor data 

**2. Lakukan observasi dengan menampilkan 10 data pertama !** 

In [3]:
# tampilkan 10 data pertama
df.head(10)

No       Species  Weight  Length1  Length2  Length3  Height  Width     Sex
0   1  Common Bream   430.0     26.5     29.0     34.0    36.6   15.1     NaN
1   2         Roach   390.0     29.5     31.7     35.0    27.1   15.3     NaN
2   3         Perch   700.0     34.0     36.0     38.3    27.7   17.6  female
3   4          Pike   430.0     35.5     38.0     40.5    18.0   11.3     NaN
4   5  Common Bream   725.0     31.8     35.0     40.9    40.0   14.8    male
5   6          Pike   300.0     31.7     34.0     37.8    15.1   11.0  female
6   7     Whitefish  1000.0     37.3     40.0     43.5    28.4   15.0     NaN
7   8         Perch   820.0     36.6     39.0     41.3    30.1   17.8     NaN
8   9         Perch   300.0     25.2     27.3     28.7    29.0   17.9  female
9  10  Common Bream   720.0     32.0     35.0     40.6    40.3   15.0     NaN

**3. Lakukan observasi dengan menampilkan 10 data terakhir !**

In [4]:
# tampilkan 10 data terakhir
df.tail(10)

No       Species  Weight  Length1  Length2  Length3  Height  Width  \
149  150  Silver Bream   300.0     24.0     26.0     29.0    39.2   14.6   
150  151          Pike  1550.0     56.0     60.0     64.0    15.0    9.6   
151  152  Common Bream   685.0     31.4     34.0     39.2    40.8   13.7   
152  153  Common Bream   680.0     31.8     35.0     40.6    38.1   15.1   
153  154          Pike  1250.0     52.0     56.0     59.7    17.9   11.7   
154  155  Common Bream   650.0     31.0     33.5     38.7    37.4   14.8   
155  156         Perch   180.0     23.0     25.0     26.5    24.3   13.9   
156  157  Common Bream   975.0     37.4     41.0     45.9    40.6   14.7   
157  158         Smelt    12.2     12.1     13.0     13.8    16.5    9.1   
158  159         Roach     0.0     19.0     20.5     22.8    28.4   14.7   

        Sex  
149  female  
150  female  
151     NaN  
152     NaN  
153     NaN  
154     NaN  
155  female  
156  female  
157  female  
158     NaN

Apakah ada yang aneh?

Dapat dilihat pada kolom yang pertama, ada kolom yang tidak relevan, sehingga Anda dapat menghilangkan kolom tersebut.<br/>
**5. Hilangkan kolom yang tidak relevan**


In [9]:
# Hilangkan kolom tidak relevan
df[:]['Species':'Sex']

TypeError: cannot do slice indexing on <class 'pandas.core.indexes.range.RangeIndex'> with these indexers [Species] of <class 'str'>

Selanjutnya, kita ingin mengetahui ada berapa baris di dalam data tersebut dengan menggunakan *method* `info`. 

In [ ]:
# panggil metode info


Terdapat 8 kolom dan 159 baris pada data tersebut. Metadata untuk dataset tersebut adalah:<br/>
- Species name of fish
- Weight of fish in Gram g
- Length1 is vertical length in cm
- Length2 is diagonal length in cm
- Length3 is cross length in cm
- Height in cm
- Width is diagonal width in cm
- Sex

Dari data `info` di atas, diketahui bahwa 6 kolom adalah data *continuous* sedangkan dua kolom merupakan objek. 


Lanjutkan Observasi data di atas, Apakah ada data `NaN`?<br/>

**6. Temukan data `NaN` dan hitung berapa jumlah record yg `NaN`**

Kita akan melakukan observasi terakhir yaitu menampilkan data apa saja di dalam kolom `Species`, dan `Sex`.

**7. Tampilkan seluruh data pada kolom `Species` dan `Sex` beserta jumlahnya**

## II. Data Pre-processing
Tahapan lanjutan adalah melakukan pembersihan atau organising atas data yang dimiliki.

> Namun, diketahui bahwa, terdapat *feature* (kolom) terdapat nilai `NaN`, di samping nilai yang sesungguhnya.

**8. Langkah apa yang akan Anda lakukan?**<br/>

Tuliskan Langkah-langkah Anda di bawah ini:

Jawaban Anda:




**9. Lakukan Langkah Yang Anda telah jelaskan di atas**

Diasumsikan bahwa data yang akan kita gunakan telah melalui serangkaian pembersihan data dan seterusnya. Tugas selanjutnya adalah membuat model untuk memprediksi Species berdasarkan atribut lainnya.

### Lakukan train test split
Sebelum membangun model, langkah yang harus dilakukan adalah memecah target variabel dan independent variabel. 

**10. Lakukan metode train test split**

In [ ]:
# Import train_test_split


# X merupakan dataset kecuali target variabel


# Y merupakan target variabel yaitu Species


# kita gunakan 70% atau 80% data training (silakan pilih)


# Creation of Train and Test dataset


# Creation of Train and validation dataset



Selanjutnya kita akan meng-import beberapa alat pengukuran 

In [ ]:
from sklearn import model_selection
from sklearn.metrics import accuracy_score
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix

**11. Lakukan Pembuatan model-model sesuai dengan pilihan Anda**<br/>
Perhatikan bahwa Anda diminta membuat **minimal 3 model** !

In [ ]:
# import library


# buat objek model


# fit model


# panggil fungsi predict


In [ ]:
print ('-'*40)
print ('Accuracy score:')
print (accuracy_score(y_test, _prediction))
print ('-'*40)
print ('Confusion Matrix:')
print (confusion_matrix(y_test, _prediction))
print ('-'*40)
print ('Classification Matrix:')
print (classification_report(y_test, _prediction))

Apabila akurasi model Anda masih di bawah 80%, maka Lakukan langkah *hyperparameter tuning* dengan mengubah konfigurasi default.

**12. Jelaskan hasil pada model Anda di atas**<br/>
- Apakah sudah sesuai dengan keinginan dan ekspektasi Anda?
- Model apa yang memiliki tingkat akurasi tertinggi?